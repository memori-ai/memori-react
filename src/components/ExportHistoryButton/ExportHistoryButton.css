.memori-export-history-button--panel {
    position: absolute;
    z-index: 10;
    right: 0.5rem;
    padding: var(--memori-button-padding);
    border: 1px solid var(--memori-button-border-color);
    border-radius: var(--memori-button-radius);
    margin-top: 0.5rem;
    background: #fff;
    box-shadow: var(--memori-button-box-shadow);
    color: #000;
    text-align: left;
}

.memori-export-history-button--panel button + button {
    margin-left: 0.5rem;
}

.memori-export-history-button--align-left .memori-export-history-button--panel {
    right: auto;
    left: 0.5rem;
}
