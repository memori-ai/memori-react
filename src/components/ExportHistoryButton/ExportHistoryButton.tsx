import React from 'react';
import { Memori, Message } from '@memori.ai/memori-api-client/dist/types';
import { useTranslation } from 'react-i18next';
import { Popover } from '@headlessui/react';
import Button from '../ui/Button';
import MessageIcon from '../icons/Message';
import cx from 'classnames';

export interface Props {
  history: Message[];
  memori: Memori;
  icon?: React.ReactNode;
  className?: string;
  buttonClassName?: string;
  align?: 'left' | 'right';
  disabled?: boolean;
  filename?: string;
}

const downloadFile = (text: string, filename = 'export-chat.txt') => {
  const data = new Blob([text], { type: 'text/plain' });
  const url = URL.createObjectURL(data);
  const element = document.createElement('a');
  element.setAttribute('href', url);
  element.setAttribute('download', filename);
  element.style.display = 'none';

  document.body.appendChild(element);
  element.click();
  document.body.removeChild(element);
};

const ExportHistoryButton = ({
  history,
  memori,
  className,
  buttonClassName,
  icon,
  disabled = false,
  align = 'right',
  filename,
}: Props) => {
  const { t } = useTranslation();

  const text = `${t(
    'write_and_speak.conversationStartedLabel'
  )} ${new Intl.DateTimeFormat('it', {
    dateStyle: 'short',
    timeStyle: 'short',
  }).format(new Date())}\n\n`.concat(
    history
      .map(m => `${m.fromUser ? 'YOU' : memori.name}: ${m.text}`)
      .join('\n')
  );

  return (
    <Popover
      title={t('write_and_speak.exportChatHistoryMessage')}
      className={cx('memori-export-history-button', className, {
        'memori-export-history-button--align-left': align === 'left',
      })}
    >
      {({ close }) => (
        <>
          <Popover.Button
            disabled={disabled}
            className={cx(
              'memori-button',
              'memori-button--circle',
              'memori-button--icon-only',
              buttonClassName,
              {
                'memori-button--disabled': disabled,
                'memori-button--primary': !disabled,
              }
            )}
          >
            <div className="memori-button--icon">{icon || <MessageIcon />}</div>
          </Popover.Button>

          <Popover.Panel className="memori-export-history-button--panel">
            <p>{t('write_and_speak.exportChatHistory')}</p>
            <Button
              primary
              onClick={() =>
                downloadFile(
                  text,
                  filename ||
                    `${memori.name.replace(/\W+/g, '-')}-export-chat.txt`
                )
              }
            >
              {t('yes') || 'Yes'}
            </Button>
            <Button onClick={() => close()}>{t('no') || 'No'}</Button>
          </Popover.Panel>
        </>
      )}
    </Popover>
  );
};

export default ExportHistoryButton;
