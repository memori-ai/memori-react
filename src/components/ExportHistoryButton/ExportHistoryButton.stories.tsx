import React from 'react';
import { Meta, Story } from '@storybook/react';
import ExportHistoryButton, { Props } from './ExportHistoryButton';
import I18nWrapper from '../../I18nWrapper';
import { memori, history } from '../../mocks/data';
import Download from '../icons/Download';

const meta: Meta = {
  title: 'Export History Button',
  component: ExportHistoryButton,
  argTypes: {
    memori: {
      control: {
        type: 'object',
      },
    },
    icon: {
      control: {
        type: 'object',
      },
    },
  },
  parameters: {
    controls: { expanded: true },
  },
};

export default meta;

const Template: Story<Props> = args => (
  <I18nWrapper>
    <div style={{ textAlign: 'right' }}>
      <ExportHistoryButton {...args} />
    </div>
  </I18nWrapper>
);

const TemplateRight: Story<Props> = args => (
  <I18nWrapper>
    <div style={{ textAlign: 'left' }}>
      <ExportHistoryButton {...args} />
    </div>
  </I18nWrapper>
);

// By passing using the Args format for exported stories, you can control the props for a component for reuse in a test
// https://storybook.js.org/docs/react/workflows/unit-testing
export const Default = Template.bind({});
Default.args = {
  memori,
  history,
};

export const WithCustomIcon = Template.bind({});
WithCustomIcon.args = {
  memori,
  history,
  icon: <Download />,
};

export const WithCustomCSSClass = Template.bind({});
WithCustomCSSClass.args = {
  memori,
  history,
  buttonClassName: 'memori-button--primary',
};

export const AlignLeft = TemplateRight.bind({});
AlignLeft.args = {
  memori,
  history,
  align: 'left',
};
