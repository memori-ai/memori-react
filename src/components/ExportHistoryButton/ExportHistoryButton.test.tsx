import React from 'react';
import { render } from '@testing-library/react';
import ExportHistoryButton from './ExportHistoryButton';
import { memori, history } from '../../mocks/data';
import Download from '../icons/Download';

import './ExportHistoryButton.css';

it('renders ExportHistoryButton unchanged', () => {
  const { container } = render(
    <ExportHistoryButton memori={memori} history={[]} />
  );
  expect(container).toMatchSnapshot();
});

it('renders ExportHistoryButton unchanged with history', () => {
  const { container } = render(
    <ExportHistoryButton memori={memori} history={history} />
  );
  expect(container).toMatchSnapshot();
});

it('renders ExportHistoryButton unchanged with className', () => {
  const { container } = render(
    <ExportHistoryButton memori={memori} history={[]} className="test" />
  );
  expect(container).toMatchSnapshot();
});

it('renders ExportHistoryButton unchanged disabled', () => {
  const { container } = render(
    <ExportHistoryButton memori={memori} history={[]} disabled />
  );
  expect(container).toMatchSnapshot();
});

it('renders ExportHistoryButton unchanged with custom icon', () => {
  const { container } = render(
    <ExportHistoryButton
      memori={memori}
      history={history}
      icon={<Download />}
    />
  );
  expect(container).toMatchSnapshot();
});

it('renders ExportHistoryButton unchanged with custom button props', () => {
  const { container } = render(
    <ExportHistoryButton
      memori={memori}
      history={history}
      icon={<Download />}
      buttonClassName="test"
    />
  );
  expect(container).toMatchSnapshot();
});

it('renders ExportHistoryButton unchanged with custom filename', () => {
  const { container } = render(
    <ExportHistoryButton
      memori={memori}
      history={[]}
      filename="awanagana.txt"
    />
  );
  expect(container).toMatchSnapshot();
});
