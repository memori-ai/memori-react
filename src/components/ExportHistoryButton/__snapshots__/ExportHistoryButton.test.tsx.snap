// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ExportHistoryButton unchanged 1`] = `
<div>
  <div
    class="memori-export-history-button"
    data-headlessui-state=""
    title="write_and_speak.exportChatHistoryMessage"
  >
    <button
      aria-expanded="false"
      class="memori-button memori-button--circle memori-button--icon-only memori-button--primary"
      data-headlessui-state=""
      id="headlessui-popover-button-:r0:"
      type="button"
    >
      <div
        class="memori-button--icon"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          role="img"
          viewBox="0 0 1024 1024"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M464 512a48 48 0 1096 0 48 48 0 10-96 0zm200 0a48 48 0 1096 0 48 48 0 10-96 0zm-400 0a48 48 0 1096 0 48 48 0 10-96 0zm661.2-173.6c-22.6-53.7-55-101.9-96.3-143.3a444.35 444.35 0 00-143.3-96.3C630.6 75.7 572.2 64 512 64h-2c-60.6.3-119.3 12.3-174.5 35.9a445.35 445.35 0 00-142 96.5c-40.9 41.3-73 89.3-95.2 142.8-23 55.4-34.6 114.3-34.3 174.9A449.4 449.4 0 00112 714v152a46 46 0 0046 46h152.1A449.4 449.4 0 00510 960h2.1c59.9 0 118-11.6 172.7-34.3a444.48 444.48 0 00142.8-95.2c41.3-40.9 73.8-88.7 96.5-142 23.6-55.2 35.6-113.9 35.9-174.5.3-60.9-11.5-120-34.8-175.6zm-151.1 438C704 845.8 611 884 512 884h-1.7c-60.3-.3-120.2-15.3-173.1-43.5l-8.4-4.5H188V695.2l-4.5-8.4C155.3 633.9 140.3 574 140 513.7c-.4-99.7 37.7-193.3 107.6-263.8 69.8-70.5 163.1-109.5 262.8-109.9h1.7c50 0 98.5 9.7 144.2 28.9 44.6 18.7 84.6 45.6 119 80 34.3 34.3 61.3 74.4 80 119 19.4 46.2 29.1 95.2 28.9 145.8-.6 99.6-39.7 192.9-110.1 262.7z"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`renders ExportHistoryButton unchanged disabled 1`] = `
<div>
  <div
    class="memori-export-history-button"
    data-headlessui-state=""
    title="write_and_speak.exportChatHistoryMessage"
  >
    <button
      class="memori-button memori-button--circle memori-button--icon-only memori-button--disabled"
      data-headlessui-state=""
      disabled=""
      id="headlessui-popover-button-:rf:"
      type="button"
    >
      <div
        class="memori-button--icon"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          role="img"
          viewBox="0 0 1024 1024"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M464 512a48 48 0 1096 0 48 48 0 10-96 0zm200 0a48 48 0 1096 0 48 48 0 10-96 0zm-400 0a48 48 0 1096 0 48 48 0 10-96 0zm661.2-173.6c-22.6-53.7-55-101.9-96.3-143.3a444.35 444.35 0 00-143.3-96.3C630.6 75.7 572.2 64 512 64h-2c-60.6.3-119.3 12.3-174.5 35.9a445.35 445.35 0 00-142 96.5c-40.9 41.3-73 89.3-95.2 142.8-23 55.4-34.6 114.3-34.3 174.9A449.4 449.4 0 00112 714v152a46 46 0 0046 46h152.1A449.4 449.4 0 00510 960h2.1c59.9 0 118-11.6 172.7-34.3a444.48 444.48 0 00142.8-95.2c41.3-40.9 73.8-88.7 96.5-142 23.6-55.2 35.6-113.9 35.9-174.5.3-60.9-11.5-120-34.8-175.6zm-151.1 438C704 845.8 611 884 512 884h-1.7c-60.3-.3-120.2-15.3-173.1-43.5l-8.4-4.5H188V695.2l-4.5-8.4C155.3 633.9 140.3 574 140 513.7c-.4-99.7 37.7-193.3 107.6-263.8 69.8-70.5 163.1-109.5 262.8-109.9h1.7c50 0 98.5 9.7 144.2 28.9 44.6 18.7 84.6 45.6 119 80 34.3 34.3 61.3 74.4 80 119 19.4 46.2 29.1 95.2 28.9 145.8-.6 99.6-39.7 192.9-110.1 262.7z"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`renders ExportHistoryButton unchanged with className 1`] = `
<div>
  <div
    class="memori-export-history-button test"
    data-headlessui-state=""
    title="write_and_speak.exportChatHistoryMessage"
  >
    <button
      aria-expanded="false"
      class="memori-button memori-button--circle memori-button--icon-only memori-button--primary"
      data-headlessui-state=""
      id="headlessui-popover-button-:ra:"
      type="button"
    >
      <div
        class="memori-button--icon"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          role="img"
          viewBox="0 0 1024 1024"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M464 512a48 48 0 1096 0 48 48 0 10-96 0zm200 0a48 48 0 1096 0 48 48 0 10-96 0zm-400 0a48 48 0 1096 0 48 48 0 10-96 0zm661.2-173.6c-22.6-53.7-55-101.9-96.3-143.3a444.35 444.35 0 00-143.3-96.3C630.6 75.7 572.2 64 512 64h-2c-60.6.3-119.3 12.3-174.5 35.9a445.35 445.35 0 00-142 96.5c-40.9 41.3-73 89.3-95.2 142.8-23 55.4-34.6 114.3-34.3 174.9A449.4 449.4 0 00112 714v152a46 46 0 0046 46h152.1A449.4 449.4 0 00510 960h2.1c59.9 0 118-11.6 172.7-34.3a444.48 444.48 0 00142.8-95.2c41.3-40.9 73.8-88.7 96.5-142 23.6-55.2 35.6-113.9 35.9-174.5.3-60.9-11.5-120-34.8-175.6zm-151.1 438C704 845.8 611 884 512 884h-1.7c-60.3-.3-120.2-15.3-173.1-43.5l-8.4-4.5H188V695.2l-4.5-8.4C155.3 633.9 140.3 574 140 513.7c-.4-99.7 37.7-193.3 107.6-263.8 69.8-70.5 163.1-109.5 262.8-109.9h1.7c50 0 98.5 9.7 144.2 28.9 44.6 18.7 84.6 45.6 119 80 34.3 34.3 61.3 74.4 80 119 19.4 46.2 29.1 95.2 28.9 145.8-.6 99.6-39.7 192.9-110.1 262.7z"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`renders ExportHistoryButton unchanged with custom button props 1`] = `
<div>
  <div
    class="memori-export-history-button"
    data-headlessui-state=""
    title="write_and_speak.exportChatHistoryMessage"
  >
    <button
      aria-expanded="false"
      class="memori-button memori-button--circle memori-button--icon-only test memori-button--primary"
      data-headlessui-state=""
      id="headlessui-popover-button-:rp:"
      type="button"
    >
      <div
        class="memori-button--icon"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          role="img"
          viewBox="0 0 1024 1024"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505.7 661a8 8 0 0012.6 0l112-141.7c4.1-5.2.4-12.9-6.3-12.9h-74.1V168c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v338.3H400c-6.7 0-10.4 7.7-6.3 12.9l112 141.8zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`renders ExportHistoryButton unchanged with custom filename 1`] = `
<div>
  <div
    class="memori-export-history-button"
    data-headlessui-state=""
    title="write_and_speak.exportChatHistoryMessage"
  >
    <button
      aria-expanded="false"
      class="memori-button memori-button--circle memori-button--icon-only memori-button--primary"
      data-headlessui-state=""
      id="headlessui-popover-button-:ru:"
      type="button"
    >
      <div
        class="memori-button--icon"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          role="img"
          viewBox="0 0 1024 1024"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M464 512a48 48 0 1096 0 48 48 0 10-96 0zm200 0a48 48 0 1096 0 48 48 0 10-96 0zm-400 0a48 48 0 1096 0 48 48 0 10-96 0zm661.2-173.6c-22.6-53.7-55-101.9-96.3-143.3a444.35 444.35 0 00-143.3-96.3C630.6 75.7 572.2 64 512 64h-2c-60.6.3-119.3 12.3-174.5 35.9a445.35 445.35 0 00-142 96.5c-40.9 41.3-73 89.3-95.2 142.8-23 55.4-34.6 114.3-34.3 174.9A449.4 449.4 0 00112 714v152a46 46 0 0046 46h152.1A449.4 449.4 0 00510 960h2.1c59.9 0 118-11.6 172.7-34.3a444.48 444.48 0 00142.8-95.2c41.3-40.9 73.8-88.7 96.5-142 23.6-55.2 35.6-113.9 35.9-174.5.3-60.9-11.5-120-34.8-175.6zm-151.1 438C704 845.8 611 884 512 884h-1.7c-60.3-.3-120.2-15.3-173.1-43.5l-8.4-4.5H188V695.2l-4.5-8.4C155.3 633.9 140.3 574 140 513.7c-.4-99.7 37.7-193.3 107.6-263.8 69.8-70.5 163.1-109.5 262.8-109.9h1.7c50 0 98.5 9.7 144.2 28.9 44.6 18.7 84.6 45.6 119 80 34.3 34.3 61.3 74.4 80 119 19.4 46.2 29.1 95.2 28.9 145.8-.6 99.6-39.7 192.9-110.1 262.7z"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`renders ExportHistoryButton unchanged with custom icon 1`] = `
<div>
  <div
    class="memori-export-history-button"
    data-headlessui-state=""
    title="write_and_speak.exportChatHistoryMessage"
  >
    <button
      aria-expanded="false"
      class="memori-button memori-button--circle memori-button--icon-only memori-button--primary"
      data-headlessui-state=""
      id="headlessui-popover-button-:rk:"
      type="button"
    >
      <div
        class="memori-button--icon"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          role="img"
          viewBox="0 0 1024 1024"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505.7 661a8 8 0 0012.6 0l112-141.7c4.1-5.2.4-12.9-6.3-12.9h-74.1V168c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v338.3H400c-6.7 0-10.4 7.7-6.3 12.9l112 141.8zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`renders ExportHistoryButton unchanged with history 1`] = `
<div>
  <div
    class="memori-export-history-button"
    data-headlessui-state=""
    title="write_and_speak.exportChatHistoryMessage"
  >
    <button
      aria-expanded="false"
      class="memori-button memori-button--circle memori-button--icon-only memori-button--primary"
      data-headlessui-state=""
      id="headlessui-popover-button-:r5:"
      type="button"
    >
      <div
        class="memori-button--icon"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          role="img"
          viewBox="0 0 1024 1024"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M464 512a48 48 0 1096 0 48 48 0 10-96 0zm200 0a48 48 0 1096 0 48 48 0 10-96 0zm-400 0a48 48 0 1096 0 48 48 0 10-96 0zm661.2-173.6c-22.6-53.7-55-101.9-96.3-143.3a444.35 444.35 0 00-143.3-96.3C630.6 75.7 572.2 64 512 64h-2c-60.6.3-119.3 12.3-174.5 35.9a445.35 445.35 0 00-142 96.5c-40.9 41.3-73 89.3-95.2 142.8-23 55.4-34.6 114.3-34.3 174.9A449.4 449.4 0 00112 714v152a46 46 0 0046 46h152.1A449.4 449.4 0 00510 960h2.1c59.9 0 118-11.6 172.7-34.3a444.48 444.48 0 00142.8-95.2c41.3-40.9 73.8-88.7 96.5-142 23.6-55.2 35.6-113.9 35.9-174.5.3-60.9-11.5-120-34.8-175.6zm-151.1 438C704 845.8 611 884 512 884h-1.7c-60.3-.3-120.2-15.3-173.1-43.5l-8.4-4.5H188V695.2l-4.5-8.4C155.3 633.9 140.3 574 140 513.7c-.4-99.7 37.7-193.3 107.6-263.8 69.8-70.5 163.1-109.5 262.8-109.9h1.7c50 0 98.5 9.7 144.2 28.9 44.6 18.7 84.6 45.6 119 80 34.3 34.3 61.3 74.4 80 119 19.4 46.2 29.1 95.2 28.9 145.8-.6 99.6-39.7 192.9-110.1 262.7z"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;
